ext/rugged/rugged_tree.c

```diff
@@ -93,6 +93,63 @@ static VALUE rb_git_tree_entrycount(VALUE self)
 	return INT2FIX(git_tree_entrycount(tree));
 }
 
+static int rugged__treecount_cb(const char *root, const git_tree_entry *entry, void *data)
+{
+	struct rugged_treecount_cb_payload *payload = data;
+
+	if (payload->limit >= 0 && payload->count >= payload->limit)
+		return -1;
+	else if(git_tree_entry_type(entry) == GIT_OBJ_TREE)
+		return 0;
+	else {
+		++(payload->count);
+		return 1;
+	}
+}
```

Comment:
Unfortunately in the libgit2 `git_tree_walk` API, there is no maximum depth check and the implementation is recursive, so there are some theoretical DOS vectors here. I think probably the best bet is to fix it at the libgit2 level and either reimplement `git_tree_walk` using the `git_iterator_for_tree` API or simply add a depth check internal to the API. (It would be nice to reflect the depth as a parameter to the callback, but that would be a breaking API change, so probably not worth it.)